backend/src/apis/round.rs

```diff
@@ -0,0 +1,226 @@
+    pub fn get_public_inputs(&self) -> &Vec<Vec<Fp>> {
+        &self.public_inputs
+    }
+
+    pub fn get_proof(&self) -> &Vec<u8> {
+        &self.proof
+    }
+}
+
+pub struct Snapshot<const LEVELS: usize, const N_ASSETS: usize, const N_BYTES: usize> {
+    mst: MerkleSumTree<N_ASSETS, N_BYTES>,
+    timestamp: usize,
+    trusted_setup: [SetupArtifacts; 2],
+}
+
+pub struct Round<const LEVELS: usize, const N_ASSETS: usize, const N_BYTES: usize> {
+    snapshot: Option<Snapshot<LEVELS, N_ASSETS, N_BYTES>>,
+    signer: SummaSigner,
+}
+
+impl<const LEVELS: usize, const N_ASSETS: usize, const N_BYTES: usize>
+    Round<LEVELS, N_ASSETS, N_BYTES>
+where
+    [usize; N_ASSETS + 1]: Sized,
+    [usize; 2 * (1 + N_ASSETS)]: Sized,
+{
+    pub fn new(
+        signer_key: &str,
+        chain_id: u64,
+        rpc_url: &str,
+        summasc_address: Address,
+    ) -> Result<Round<LEVELS, N_ASSETS, N_BYTES>, Box<dyn Error>> {
+        Ok(Round {
+            snapshot: None,
+            signer: SummaSigner::new(&vec![], signer_key, chain_id, rpc_url, summasc_address),
+        })
+    }
+
+    pub fn build_snapshot(&mut self, entry_csv_path: &str, params_path: &str, timestamp: usize) {
+        let snapshot =
+            Snapshot::<LEVELS, N_ASSETS, N_BYTES>::new(entry_csv_path, params_path, timestamp)
+                .unwrap();
+        self.snapshot = Some(snapshot);
+    }
+
+    pub async fn dispatch_solvency_proof(
+        &mut self,
+        assets: [Asset; N_ASSETS],
+    ) -> Result<(), &'static str> {
+        if self.snapshot.is_none() {
+            return Err("snapshot is not built yet");
+        }
+        let snapshot = self.snapshot.as_ref().unwrap();
+
+        // Convert U256 to Fp for generating proof of solvency
+        let asset_sum: [Fp; N_ASSETS] = assets
+            .iter()
+            .map(|asset| Fp::from_raw(asset.amount.0) as Fp)
+            .collect::<Vec<Fp>>()
+            .try_into()
+            .unwrap();
+
+        let proof: SolvencyProof = match snapshot.generate_proof_of_solvency(asset_sum) {
+            Ok(p) => p,
+            Err(_) => return Err("Failed to generate proof of solvency"),
+        };
+
+        let result = self
+            .signer
+            .submit_proof_of_solvency(
+                proof.public_inputs[0],
+                assets.to_vec(),
+                proof.proof_calldata,
+                U256::from(snapshot.get_timestamp()),
+            )
+            .await;
+
+        Ok(result.unwrap())
+    }
+
+    pub fn get_proof_of_inclusion(
+        &self,
+        user_index: usize,
+    ) -> Result<MstInclusionProof, &'static str> {
+        let snapshot = self.snapshot.as_ref().unwrap();
+        if snapshot.mst.entries().len() < user_index {
+            return Err("user_index is out of range");
+        }
+
+        Ok(snapshot.generate_proof_of_inclusion(user_index).unwrap())
+    }
+}
+
+impl<const LEVELS: usize, const N_ASSETS: usize, const N_BYTES: usize>
```

Comment:
I believe that Snapshot doesn't have to be exposed to the public since the exchange would only potentially interact with `Round`, therefore I think that we can safely remove `pub` from these function. I'm not sure about that tho

Reply:
Currently, the snapshot is directly used in the inclusion verifier example, which is why I've made it public. On the issue #128 I'll consider making it private if it's appropriate.